macos/library/FLEViewController.m

```diff
@@ -265,29 +285,43 @@ + (FlutterRendererConfig)createRenderConfigHeadless:(BOOL)headless {
 }
 
 - (void)handlePlatformMessage:(const FlutterPlatformMessage *)message {
-  // This method must not return early as we still need to submit the
-  // response callback in case of error.
-
   NSData *data = [NSData dataWithBytesNoCopy:(void *)message->message
                                       length:message->message_size
                                 freeWhenDone:NO];
-
-  NSError *error = nil;
-  NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
-
   NSString *channel = @(message->channel);
-  id<FLEPlugin> plugin = _plugins[channel];
-  if (plugin != nil && dictionary != nil) {
-    id response = [plugin handlePlatformMessage:dictionary];
 
-    NSData *responseData = nil;
-    if (response != nil) {
-      responseData = [NSJSONSerialization dataWithJSONObject:response options:0 error:&error];
+  id messageObject = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
+  FLEMethodCall *methodCall =
+      messageObject ? [FLEMethodCall methodCallFromMessage:messageObject] : nil;
+
+  __block const FlutterPlatformMessageResponseHandle *responseHandle = message->response_handle;
+  FLEMethodResult resultHandler = ^(id result) {
+    NSData *responseData = EngineResponseForMethodResult(result);
+    if (responseHandle) {
+      FlutterEngineSendPlatformMessageResponse(_engine, responseHandle, responseData.bytes,
+                                               responseData.length);
+      responseHandle = NULL;
+    } else {
+      NSLog(
+          @"Error: Method call responses can be called only once. Ignoring duplicate response "
+           "for '%@' on channel '%@'.",
+          methodCall.methodName, channel);
     }
+  };
 
-    FlutterEngineSendPlatformMessageResponse(_engine, message->response_handle, responseData.bytes,
-                                             responseData.length);
+  if (!methodCall) {
+    NSLog(@"Recieved invalid platform message: %@", messageObject);
+    resultHandler(FLEMethodNotImplemented);
+    return;
   }
+
+  id<FLEPlugin> plugin = _plugins[channel];
+  if (!plugin) {
+    resultHandler(FLEMethodNotImplemented);
```

Comment:
It seems like this would warrant a more specific error than FLEMethodNotImplemented?  Is this the same in the iOS implementation?

Reply:
> Is this the same in the iOS implementation?

I believe so, yes:
https://github.com/flutter/engine/blob/58e84c8bf0b6304649340e6eaad988602961d7f6/shell/platform/darwin/ios/framework/Source/platform_message_router.mm#L38
An empty reply is what the higher-level code sends for FLEMethodNotImplemented.